app/helpers/visns_helper.rb

```diff
@@ -1,35 +1,144 @@
           end
         end
+      # add practices that have an initiating_facility
       elsif p.visn? && initiating_facility.present?
         visn_created_practices << { "visn": initiating_facility } if initiating_facility === visn.id.to_s
       end
     end
     visn_created_practices.count
   end
+
+  def get_facility_locations_by_visn(visn)
+    visn_facility_locations = []
+
+    # add facility locations to empty array
+    visn.get_va_facilities.each do |vaf|
+      facility_location = vaf.street_address_state
+
+      visn_facility_locations << facility_location unless visn_facility_locations.include?(facility_location)
+    end
+
+    location_list = ''
+    # sort the locations alphabetically
+    sorted_facility_locations = visn_facility_locations.sort
+
+    # Add other US territories to us_states helper method array
+    va_facility_locations = us_states.concat(
+      [
+       ["Virgin Islands", "VI"],
+       ["Philippines Islands", "PI"],
+       ["Guam", "GU"],
+       ["American Samoa", "AS"]
+      ]
+    )
+
+    # iterate through the facility locations and add text
+    sorted_facility_locations.each do |sfl|
+      va_facility_locations.each do |vfl|
+        full_name = vfl.first === "Virgin Islands" || vfl.first === "Philippines Islands" ? "the #{vfl.first}" : vfl.first
+        if vfl[1] === sfl
+          if sorted_facility_locations.count > 1 && sorted_facility_locations.last === sfl
+            location_list += "and #{full_name}"
+          elsif sorted_facility_locations.count > 2
+            location_list += "#{full_name}, "
+          else
+            location_list += sorted_facility_locations.count == 1 ? "#{full_name}" : "#{full_name} "
+          end
+        end
+      end
+    end
+
+    location_list
+  end
+
+  def facility_type_count(facility_type_array, facility_type)
+    facility_type_array.select { |type| type === facility_type }.count
+  end
+
+  def facility_type_counts_by_visn(visn)
+    visn_facility_types_arr = []
+
+    # add facility types to empty array
+    visn.get_va_facilities.each do |vaf|
+      facility_type = vaf.classification
+
+      visn_facility_types_arr << facility_type
+    end
+
+    # return the counts for each facility type
+    [
+      facility_type_count(visn_facility_types_arr, 'Health Care Center (HCC)'),
+      facility_type_count(visn_facility_types_arr, 'Multi-Specialty CBOC'),
+      facility_type_count(visn_facility_types_arr, 'Other Outpatient Services (OOS)'),
+      facility_type_count(visn_facility_types_arr, 'Primary Care CBOC'),
+      facility_type_count(visn_facility_types_arr, 'Residential Care Site (MH RRTP/DRRTP) (Stand-Alone)'),
+      facility_type_count(visn_facility_types_arr, 'Unclassified'),
+      facility_type_count(visn_facility_types_arr, 'VA Medical Center (VAMC)')
+    ]
+  end
+
+  def get_facility_type_text_by_visn(visn)
+    health_care_center_count = facility_type_counts_by_visn(visn)[0]
```

Comment:
could we just set the value of `facility_type_counts_by_visn` to a variable and then call that variable when we are assigning the counts so we don't have to call the same method multiple times?

Reply:
refactored